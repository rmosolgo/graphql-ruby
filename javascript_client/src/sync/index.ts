import sendPayload from "./sendPayload"
import { generateClientCode, gatherOperations, ClientOperation } from "./generateClient"
import Logger from "./logger"
import fs from "fs"

interface SyncOptions {
  path?: string,
  relayPersistedOutput?: string,
  apolloAndroidOperationOutput?: string,
  secret?: string
  url?: string,
  mode?: string,
  outfile?: string,
  outfileType?: string,
  client: string,
  send?: Function,
  hash?: Function,
  verbose?: boolean,
  quiet?: boolean,
  addTypename?: boolean,
}
/**
 * Find `.graphql` files in `path`,
 * then prepare them & send them to the configured endpoint.
 *
 * @param {Object} options
 * @param {String} options.path - A glob to recursively search for `.graphql` files (Default is `./`)
 * @param {String} options.relayPersistedOutput - A path to a `.json` file from `relay-compiler`'s  `--persist-output` option
 * @param {String} options.secret - HMAC-SHA256 key which must match the server secret (default is no encryption)
 * @param {String} options.url - Target URL for sending prepared queries. If omitted, then an outfile is generated without sending operations to the server.
 * @param {String} options.mode - If `"file"`, treat each file separately. If `"project"`, concatenate all files and extract each operation. If `"relay"`, treat it as relay-compiler output
 * @param {Boolean} options.addTypename - Indicates if the "__typename" field are automatically added to your queries
 * @param {String} options.outfile - Where the generated code should be written
 * @param {String} options.outfileType - The type of the generated code (i.e., json, js)
 * @param {String} options.client - the Client ID that these operations belong to
 * @param {Function} options.send - A function for sending the payload to the server, with the signature `options.send(payload)`. (Default is an HTTP `POST` request)
 * @param {Function} options.hash - A custom hash function for query strings with the signature `options.hash(string) => digest` (Default is `md5(string) => digest`)
 * @param {Boolean} options.verbose - If true, log debug output
 * @return {Promise} Rejects with an Error or String if something goes wrong. Resolves with the operation payload if successful.
*/
function sync(options: SyncOptions) {
  var logger = new Logger(!!options.quiet)
  var verbose = !!options.verbose
  var url = options.url
  if (!url) {
    logger.log("No URL; Generating artifacts without syncing them")
  }
  var clientName = options.client
  if (!clientName) {
    throw new Error("Client name must be provided for sync")
  }
  var encryptionKey = options.secret
  if (encryptionKey) {
    logger.log("Authenticating with HMAC")
  }

  var graphqlGlob = options.path
  var hashFunc = options.hash
  var sendFunc = options.send || sendPayload
  var gatherMode = options.mode
  var clientType = options.outfileType
  if (options.relayPersistedOutput) {
    // relay-compiler has already generated an artifact for us
    var payload: { operations: ClientOperation[] } = { operations: [] }
    var relayOutputText = fs.readFileSync(options.relayPersistedOutput, "utf8")
    var relayOutput = JSON.parse(relayOutputText)
    var operationBody
    for (var hash in relayOutput) {
      operationBody = relayOutput[hash]
      payload.operations.push({
        body: operationBody,
        alias: hash,
      })
    }
  } else if (options.apolloAndroidOperationOutput) {
    // Apollo Android has already generated an artifact (https://www.apollographql.com/docs/android/advanced/persisted-queries/#operationoutputjson)
    var payload: { operations: ClientOperation[] } = { operations: [] }
    var apolloAndroidOutputText = fs.readFileSync(options.apolloAndroidOperationOutput, "utf8")
    var apolloAndroidOutput = JSON.parse(apolloAndroidOutputText)
    var operationData
    // Structure is { operationId => { "name" => "...", "source" => "query { ... } " } }
    for (var operationId in apolloAndroidOutput) {
      operationData = apolloAndroidOutput[operationId]
      payload.operations.push({
        body: operationData.source,
        alias: operationId,
      })
    }

  } else {
    var payload = gatherOperations({
      path: graphqlGlob,
      hash: hashFunc,
      mode: gatherMode,
      addTypename: !!options.addTypename,
      clientType: clientType,
      client: clientName,
      verbose: verbose,
    })
  }

  var outfile: string | null
  if (options.outfile) {
    outfile = options.outfile
  } else if (options.relayPersistedOutput || options.apolloAndroidOperationOutput) {
    // These artifacts have embedded IDs in its generated files,
    // no need to generate an outfile.
    outfile = null
  } else if (fs.existsSync("src")) {
    outfile = "src/OperationStoreClient.js"
  } else {
    outfile = "OperationStoreClient.js"
  }

  var syncPromise = new Promise(function(resolve, reject) {
    if (payload.operations.length === 0) {
      logger.log("No operations found in " + options.path + ", not syncing anything")
      resolve(null)
      return
    } else if(!url) {
      // This is a local-only run to generate an artifact
      resolve(payload)
      return
    } else {
      logger.log("Syncing " + payload.operations.length + " operations to " + logger.bright(url) + "...")
      var sendOpts = {
        url: url,
        client: clientName,
        secret: encryptionKey,
        verbose: verbose,
      }
      var sendPromise = Promise.resolve(sendFunc(payload, sendOpts))
      return sendPromise.then(function(response) {
        var responseData
        if (response) {
          try {
            responseData = JSON.parse(response)
            var aliasToNameMap: {[key: string] : string | undefined} = {}

            payload.operations.forEach(function(op) {
              aliasToNameMap[op.alias] = op.name
            })

            var failed = responseData.failed.length
            // These might get overriden for status output
            var notModified = responseData.not_modified.length
            var added = responseData.added.length
            if (failed) {
              // Override these to reflect reality
              notModified = 0
              added = 0
            }

            var addedColor = added ? "green" : "dim"
            logger.log("  " + logger.colorize(addedColor, added + " added"))
            var notModifiedColor = notModified ? "reset" : "dim"

            logger.log("  " + logger.colorize(notModifiedColor, notModified + " not modified"))
            var failedColor = failed ? "red" : "dim"
            logger.log("  " + logger.colorize(failedColor, failed + " failed"))

            if (failed) {
              logger.error("Sync failed, errors:")
              var failedOperationAlias: string
              var failedOperationName: string
              var errors
              var allErrors: string[] = []
              for (failedOperationAlias in responseData.errors) {
                failedOperationName = aliasToNameMap[failedOperationAlias] || failedOperationAlias
                logger.error("  " + failedOperationName + ":")
                errors = responseData.errors[failedOperationAlias]
                errors.forEach(function(errMessage: string) {
                  allErrors.push(failedOperationName + ": " + errMessage)
                  logger.error("    " + logger.red("✘") + " " + errMessage)
                })
              }
              reject("Sync failed: " + allErrors.join(", "))
              return
            }
          } catch (err) {
            logger.log("Failed to print sync result:", err)
            reject(err)
            return
          }
        }
        resolve(payload)
        return
      }).catch(function(err) {
        logger.error(logger.red("Sync failed:"))
        logger.error(err)
        reject(err)
        return
      })
    }
  })

  return syncPromise.then(function(_payload) {
    // The payload is yielded when sync was successful, but typescript had
    // trouble using it from ^^ here. So instead, just use its presence as a signal to continue.

    // Don't generate a new file when we're using relay-comipler's --persist-output
    if (_payload && outfile) {
      var generatedCode = generateClientCode(clientName, payload.operations, clientType)
      var finishedPayload = {
        operations: payload.operations,
        generatedCode,
      }
      logger.log("Generating client module in " + logger.colorize("bright", outfile) + "...")
      fs.writeFileSync(outfile, generatedCode, "utf8")
      logger.log(logger.green("✓ Done!"))
      return finishedPayload
    } else {
      logger.log(logger.green("✓ Done!"))
      return payload
    }
  })
}

export default sync
