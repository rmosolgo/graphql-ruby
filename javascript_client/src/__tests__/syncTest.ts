import sync from "../sync"
var fs = require("fs")
var nock = require("nock")

interface MockOperation {
  alias: string,
}

interface MockPayload {
  operations: MockOperation[],
  generatedCode: string,
}

interface MockedObject {
   mock: { calls: object }
}

describe("sync operations", () => {
  beforeEach(() => {
    global.console.error = jest.fn()
    global.console.log = jest.fn()
  })

  afterEach(() => {
    jest.clearAllMocks();
  })

  describe("generating artifacts without syncing", () => {
    it("works without a URL", () => {
      var options = {
        client: "test-1",
        path: "./src/__tests__/documents",
      }

      return sync(options).then(function() {
        var generatedCode = fs.readFileSync("./src/OperationStoreClient.js", "utf8")
        expect(generatedCode).toMatch('"GetStuff": "f7f65309043352183e905e1396e51078"')
        expect(generatedCode).toMatchSnapshot()
      })
    })
  })

  describe("custom HTTP options", () => {
    it("uses the provided `send` option & provided URL", () => {
      var url: string
      var options = {
        client: "test-1",
        path: "./src/__tests__/documents",
        url: "bogus",
        quiet: true,
        send: (_sendPayload: object, options: { url: string }) => {
          url = options.url
        },
      }
      return sync(options).then(function() {
        expect(url).toEqual("bogus")
      })
    })
  })

  describe("verbose", () => {
    it("Adds debug output", () => {
      var spy = (console.log as unknown) as MockedObject
      var options = {
        client: "test-1",
        path: "./src/__tests__/documents",
        url: "bogus",
        verbose: true,
        send: (_sendPayload: string, opts: { verbose: boolean }) => {
          if (opts.verbose) {
            console.log("Verbose!")
          }
        },
      }
      return sync(options).then(function() {
        expect(spy.mock.calls).toMatchSnapshot()
      })
    })
  })

  describe("custom file processing options", () => {
    it("Adds .graphql to the glob if needed", () => {
      var payload: MockPayload
      var options = {
        client: "test-1",
        path: "./src/__tests__/documents",
        url: "bogus",
        quiet: true,
        send: (sendPayload: MockPayload, _opts: object) => { payload = sendPayload },
      }
      return sync(options).then(function() {
        expect(payload.operations).toMatchSnapshot()

        var optionsWithExt = {...options, glob: "./**/*.graphql"}
        return sync(optionsWithExt).then(function() {
          // Get the same result, even when the glob already has a file extension
          expect(payload.operations).toMatchSnapshot()
        })
      })
    })

    it("Uses a custom hash function if provided", () => {
      var payload: MockPayload
      var options = {
        client: "test-1",
        path: "./src/__tests__/documents",
        url: "bogus",
        quiet: true,
        hash: (graphQLBody: string) => {
          // This is a bad hack to get the operation name
          var opName = graphQLBody.match(/query ([A-Za-z]+) \{/)
          return opName ? opName[1].toUpperCase() : null
        },
        send: (sendPayload: MockPayload, _opts: object) => { payload = sendPayload },
      }
      return sync(options).then(function() {
        expect(payload.operations).toMatchSnapshot()
      })
    })
  })

  describe("Relay support", () => {
    it("Uses Relay generated .js files", () => {
      var payload: MockPayload
      var options = {
        client: "test-1",
        quiet: true,
        path: "./src/__generated__",
        url: "bogus",
        send: (sendPayload: MockPayload, _opts: object) => { payload = sendPayload },
      }
      return sync(options).then(function () {
        expect(payload.operations).toMatchSnapshot()
      })
    })

    it("Uses relay --persist-output JSON files", () => {
      var payload: MockPayload
      var options = {
        client: "test-1",
        quiet: true,
        relayPersistedOutput: "./src/__tests__/example-relay-persisted-queries.json",
        url: "bogus",
        send: (sendPayload: MockPayload, _opts: object) => {
          payload = sendPayload
        },
      }
      return sync(options).then(function () {
        return expect(payload.operations).toMatchSnapshot()
      })
    })

    it("Uses Apollo Android OperationOutput JSON files", () => {
      var payload: MockPayload
      var options = {
        client: "test-1",
        quiet: true,
        apolloAndroidOperationOutput: "./src/__tests__/example-apollo-android-operation-output.json",
        url: "bogus",
        send: (sendPayload: MockPayload, _opts: object) => {
          payload = sendPayload
        },
      }
      return sync(options).then(function () {
        expect(payload.operations[0].alias).toEqual("aba626ea9bdf465954e89e5590eb2c1a")
        return expect(payload.operations).toMatchSnapshot()
      })
    })
  })

  describe("Input files", () => {
    it("Merges fragments and operations across files", () => {
      var payload: MockPayload
      var options = {
        client: "test-1",
        quiet: true,
        path: "./src/__tests__/project/",
        url: "bogus",
        // mode: "project" is the default
        send: (sendPayload: MockPayload, _opts: object) => { payload = sendPayload },
      }
      return sync(options).then(function () {
        expect(payload.operations).toMatchSnapshot()
      })
    })

    it("Uses mode: file to process each file separately", () => {
      var payload: MockPayload
      var options = {
        client: "test-1",
        quiet: true,
        path: "./src/__tests__/project",
        url: "bogus",
        mode: "file",
        send: (sendPayload: MockPayload, _opts: object) => { payload = sendPayload },
      }
      return sync(options).then(function() {
        expect(payload.operations).toMatchSnapshot()
      })
    })
  })

  describe("Promise result", () => {
    it("Yields the payload and generated code", () => {
      var options = {
        client: "test-1",
        path: "./src/__tests__/project",
        url: "bogus",
        quiet: true,
        send: () => { },
      }

      return sync(options).then(function(ppayload: unknown) {
        var payload = ppayload as MockPayload
        expect(payload.operations.length).toEqual(5)
        var generatedCode = fs.readFileSync("./src/OperationStoreClient.js", "utf8")
        expect(payload.generatedCode).toEqual(generatedCode)
        fs.unlinkSync("./src/OperationStoreClient.js")
      })
    })
  })
  describe("Sync output", () => {
    it("Generates a usable artifact for middleware", () => {
      var options = {
        client: "test-1",
        path: "./src/__tests__/project",
        url: "bogus",
        quiet: true,
        send: () => { },
      }
      return sync(options).then(function() {
        var generatedCode = fs.readFileSync("./src/OperationStoreClient.js", "utf8")
        expect(generatedCode).toMatch('"GetStuff": "5f0da489cf508a7c65ff5fa144e50545"')
        expect(generatedCode).toMatch('module.exports = OperationStoreClient')
        expect(generatedCode).toMatch('var _client = "test-1"')
        fs.unlinkSync("./src/OperationStoreClient.js")
      })
    })

    it("Takes an outfile option", () => {
      var options = {
        client: "test-2",
        path: "./src/__tests__/project",
        url: "bogus",
        quiet: true,
        outfile: "__crazy_outfile.js",
        send: () => { },
      }
      return sync(options).then(function() {
        var generatedCode = fs.readFileSync("./__crazy_outfile.js", "utf8")
        expect(generatedCode).toMatch('"GetStuff": "5f0da489cf508a7c65ff5fa144e50545"')
        expect(generatedCode).toMatch('module.exports = OperationStoreClient')
        expect(generatedCode).toMatch('var _client = "test-2"')
        fs.unlinkSync("./__crazy_outfile.js")
      })
    })

    it("Skips outfile generation when using --persist-output artifact", () => {
      var options = {
        client: "test-2",
        relayPersistedOutput: "./src/__tests__/example-relay-persisted-queries.json",
        url: "bogus",
        quiet: true,
        send: () => { },
      }
      return sync(options).then(function() {
        // This is the default outfile:
        var wasWritten = fs.existsSync("./src/OperationStoreClient.js")
        expect(wasWritten).toBe(false)
      })
    })

    it("Skips outfile generation when using --apollo-android-operation-output artifact", () => {
      var options = {
        client: "test-2",
        apolloAndroidOperationOutput: "./src/__tests__/example-apollo-android-operation-output.json",
        url: "bogus",
        quiet: true,
        send: () => { },
      }
      return sync(options).then(function() {
        // This is the default outfile:
        var wasWritten = fs.existsSync("./src/OperationStoreClient.js")
        expect(wasWritten).toBe(false)
      })
    })
  })

  describe("Logging", () => {
    it("Logs progress", () => {
      var spy = (console.log as unknown) as MockedObject

      var options = {
        client: "test-1",
        path: "./src/__tests__/project",
        url: "bogus",
        send: () => { },
      }
      return sync(options).then(function() {
        expect(spy.mock.calls).toMatchSnapshot()
      })
    })

    it("Can be quieted with quiet: true", () => {
      var spy = (console.log as unknown) as MockedObject

      var options = {
        client: "test-1",
        path: "./src/__tests__/project",
        url: "bogus",
        quiet: true,
        send: () => { },
      }
      return sync(options).then(function() {
        expect(spy.mock.calls).toMatchSnapshot()
      })
    })
  })

  describe("Printing the result", () => {
    function buildMockRespondingWith(status: number, data: object) {
      return nock("http://example.com").post("/stored_operations/sync").reply(status, data)
    }

    it("prints failure and sends the message to the promise", () => {
      var spyConsoleLog = (console.log as unknown) as MockedObject
      var spyConsoleError = (console.error as unknown) as MockedObject

      buildMockRespondingWith(422, {
        errors: { "5f0da489cf508a7c65ff5fa144e50545": ["something"] },
        failed: ["5f0da489cf508a7c65ff5fa144e50545"],
        added: ["defg"],
        not_modified: [],
      })

      var options = {
        client: "test-1",
        path: "./src/__tests__/project",
        url: "http://example.com/stored_operations/sync",
        quiet: false,
      }

      var syncPromise = sync(options)

      return syncPromise.catch((errmsg) => {
        expect(errmsg).toEqual("Sync failed: GetStuff: something")
        expect(spyConsoleLog.mock.calls).toMatchSnapshot()
        expect(spyConsoleError.mock.calls).toMatchSnapshot()
        jest.clearAllMocks();
      })
    })

    it("prints success", () => {
      var spyConsoleLog = (console.log as unknown) as MockedObject
      var spyConsoleError = (console.error as unknown) as MockedObject

      buildMockRespondingWith(422, {
        errors: {},
        failed: [],
        added: ["defg"],
        not_modified: ["xyz", "123"],
      })

      var options = {
        client: "test-1",
        path: "./src/__tests__/project",
        url: "http://example.com/stored_operations/sync",
        quiet: false,
      }

      var syncPromise = sync(options)

      expect(spyConsoleLog.mock.calls).toMatchSnapshot()
      jest.clearAllMocks();

      return syncPromise.then(() => {
        expect(spyConsoleLog.mock.calls).toMatchSnapshot()
        expect(spyConsoleError.mock.calls).toMatchSnapshot()
        jest.clearAllMocks();
      })
    })
  })
})
