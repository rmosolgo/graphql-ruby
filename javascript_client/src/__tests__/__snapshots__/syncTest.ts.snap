// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sync operations Input files Merges fragments and operations across files 1`] = `
Array [
  Object {
    "alias": "5f0da489cf508a7c65ff5fa144e50545",
    "body": "fragment Frag1 on Query {
  moreStuff
}

query GetStuff {
  ...Frag1
}
",
    "name": "GetStuff",
  },
  Object {
    "alias": "c944b08d15eb94cf93dd124b7d664b62",
    "body": "fragment Frag1 on Query {
  moreStuff
}

fragment Frag2 on Query {
  ...Frag3
}

fragment Frag3 on Query {
  evenMoreStuff
}

query GetStuff2 {
  stuff
  ...Frag1
  ...Frag2
}
",
    "name": "GetStuff2",
  },
  Object {
    "alias": "fc215a466a3ea309fe493e8f7cb206f2",
    "body": "fragment Frag2 on Query {
  ...Frag3
}

fragment Frag3 on Query {
  evenMoreStuff
}

fragment Frag4 on Query {
  evenMoreStuff {
    stuffInside
  }
}

query GetStuff3 {
  stuff {
    withStuffInside
  }
  ...Frag2
  ...Frag4
}
",
    "name": "GetStuff3",
  },
  Object {
    "alias": "919af8f80d8848a9b9b6176e457f8e45",
    "body": "query GetStuffIsolated {
  ...FragIsolated
  things {
    existHere
  }
}

fragment FragIsolated on Query {
  evenMoreStuff {
    stuffInside
  }
}
",
    "name": "GetStuffIsolated",
  },
  Object {
    "alias": "d5dfe825034aeda06e2639935d8b107d",
    "body": "query GetStuffIsolated2 {
  things {
    existHere
  }
}
",
    "name": "GetStuffIsolated2",
  },
]
`;

exports[`sync operations Input files Uses mode: file to process each file separately 1`] = `
Array [
  Object {
    "alias": "26c44bfe42872860da112b6177355bfa",
    "body": "fragment Frag1 on Query {
  moreStuff
}
",
    "name": "",
  },
  Object {
    "alias": "7de9e7bf1d6ea1f527de07a25983086c",
    "body": "fragment Frag2 on Query {
  ...Frag3
}
",
    "name": "",
  },
  Object {
    "alias": "8bc9b9922a7fbb66f4ac6c58d5d5c357",
    "body": "fragment Frag3 on Query {
  evenMoreStuff
}
",
    "name": "",
  },
  Object {
    "alias": "bfc521a5ade5ff6f17bdf9352c86c850",
    "body": "fragment Frag4 on Query {
  evenMoreStuff {
    stuffInside
  }
}
",
    "name": "",
  },
  Object {
    "alias": "0a6add7303775e2487f2c2235ecb1c80",
    "body": "query GetStuff {
  ...Frag1
}
",
    "name": "GetStuff",
  },
  Object {
    "alias": "2f26b770ded2a04279bc4bf824ca54ac",
    "body": "query GetStuff2 {
  stuff
  ...Frag1
  ...Frag2
}
",
    "name": "GetStuff2",
  },
  Object {
    "alias": "1b8da77aabef67ee54df7e5acfd75893",
    "body": "query GetStuff3 {
  stuff {
    withStuffInside
  }
  ...Frag2
  ...Frag4
}
",
    "name": "GetStuff3",
  },
  Object {
    "alias": "919af8f80d8848a9b9b6176e457f8e45",
    "body": "query GetStuffIsolated {
  ...FragIsolated
  things {
    existHere
  }
}

fragment FragIsolated on Query {
  evenMoreStuff {
    stuffInside
  }
}
",
    "name": "GetStuffIsolated",
  },
  Object {
    "alias": "d5dfe825034aeda06e2639935d8b107d",
    "body": "query GetStuffIsolated2 {
  things {
    existHere
  }
}
",
    "name": "GetStuffIsolated2",
  },
]
`;

exports[`sync operations Logging Can be quieted with quiet: true 1`] = `Array []`;

exports[`sync operations Logging Logs progress 1`] = `
Array [
  Array [
    "Syncing 5 operations to [1mbogus[0m...",
  ],
  Array [
    "Generating client module in [1msrc/OperationStoreClient.js[0m...",
  ],
  Array [
    "[32mâœ“ Done![0m",
  ],
]
`;

exports[`sync operations Printing the result prints failure and sends the message to the promise 1`] = `
Array [
  Array [
    "Syncing 5 operations to [1mhttp://example.com/stored_operations/sync[0m...",
  ],
  Array [
    "  [2m0 added[0m",
  ],
  Array [
    "  [2m0 not modified[0m",
  ],
  Array [
    "  [31m1 failed[0m",
  ],
]
`;

exports[`sync operations Printing the result prints failure and sends the message to the promise 2`] = `
Array [
  Array [
    "Sync failed, errors:",
  ],
  Array [
    "  GetStuff:",
  ],
  Array [
    "    [31mâœ˜[0m something",
  ],
]
`;

exports[`sync operations Printing the result prints success 1`] = `
Array [
  Array [
    "Syncing 5 operations to [1mhttp://example.com/stored_operations/sync[0m...",
  ],
]
`;

exports[`sync operations Printing the result prints success 2`] = `
Array [
  Array [
    "  [32m1 added[0m",
  ],
  Array [
    "  [0m2 not modified[0m",
  ],
  Array [
    "  [2m0 failed[0m",
  ],
  Array [
    "Generating client module in [1msrc/OperationStoreClient.js[0m...",
  ],
  Array [
    "[32mâœ“ Done![0m",
  ],
]
`;

exports[`sync operations Printing the result prints success 3`] = `Array []`;

exports[`sync operations Relay support Uses Apollo Android OperationOutput JSON files 1`] = `
Array [
  Object {
    "alias": "aba626ea9bdf465954e89e5590eb2c1a",
    "body": "mutation RemoveTodoMutation(
  $input: RemoveTodoInput!
) {
  removeTodo(input: $input) {
    deletedTodoId
    user {
      completedCount
      totalCount
      id
    }
  }
}
",
  },
  Object {
    "alias": "67c2bc8aa3185a209d6651b4feb63c04",
    "body": "query appQuery(
  $userId: String
) {
  user(id: $userId) {
    ...TodoApp_user
    id
  }
}

fragment TodoApp_user on User {
  id
  userId
  totalCount
  ...TodoListFooter_user
  ...TodoList_user
}

fragment TodoListFooter_user on User {
  id
  userId
  completedCount
  todos(first: 2147483647) {
    edges {
      node {
        id
        complete
        __typename
      }
      cursor
    }
    pageInfo {
      endCursor
      hasNextPage
    }
  }
  totalCount
}

fragment TodoList_user on User {
  todos(first: 2147483647) {
    edges {
      node {
        id
        complete
        ...Todo_todo
        __typename
      }
      cursor
    }
    pageInfo {
      endCursor
      hasNextPage
    }
  }
  id
  userId
  totalCount
  completedCount
  ...Todo_user
}

fragment Todo_todo on Todo {
  complete
  id
  text
}

fragment Todo_user on User {
  id
  userId
  totalCount
  completedCount
}
",
  },
  Object {
    "alias": "db9904c31d91416f21d45fe3d153884c",
    "body": "mutation MarkAllTodosMutation(
  $input: MarkAllTodosInput!
) {
  markAllTodos(input: $input) {
    changedTodos {
      id
      complete
    }
    user {
      id
      completedCount
    }
  }
}
",
  },
  Object {
    "alias": "2eb8c9941fdb3117fdbc08d15fab62d0",
    "body": "mutation AddTodoMutation(
  $input: AddTodoInput!
) {
  addTodo(input: $input) {
    todoEdge {
      __typename
      cursor
      node {
        complete
        id
        text
      }
    }
    user {
      id
      totalCount
    }
  }
}
",
  },
  Object {
    "alias": "d970fd7dbf118794415dec7324d463e3",
    "body": "mutation RenameTodoMutation(
  $input: RenameTodoInput!
) {
  renameTodo(input: $input) {
    todo {
      id
      text
    }
  }
}
",
  },
  Object {
    "alias": "a49217db31a8be3f4107763b957d5fca",
    "body": "mutation RemoveCompletedTodosMutation(
  $input: RemoveCompletedTodosInput!
) {
  removeCompletedTodos(input: $input) {
    deletedTodoIds
    user {
      completedCount
      totalCount
      id
    }
  }
}
",
  },
  Object {
    "alias": "d7dda774dcfa32fe0d9661e01cac9a4a",
    "body": "mutation ChangeTodoStatusMutation(
  $input: ChangeTodoStatusInput!
) {
  changeTodoStatus(input: $input) {
    todo {
      id
      complete
    }
    user {
      id
      completedCount
    }
  }
}
",
  },
]
`;

exports[`sync operations Relay support Uses Relay generated .js files 1`] = `
Array [
  Object {
    "alias": "353e010cb78d082b29cb63ee7e9027b3",
    "body": "query AppFeedQuery {
  feed(type: NEW, limit: 5) {
    ...Feed
  }
}

fragment Feed on Entry {
  repository {
    owner {
      login
    }
    name
  }
  ...FeedEntry
}

fragment FeedEntry on Entry {
  repository {
    owner {
      login
    }
    name
    stargazers_count
  }
  postedBy {
    login
  }
}
",
    "name": "AppFeedQuery",
  },
]
`;

exports[`sync operations Relay support Uses relay --persist-output JSON files 1`] = `
Array [
  Object {
    "alias": "aba626ea9bdf465954e89e5590eb2c1a",
    "body": "mutation RemoveTodoMutation(
  $input: RemoveTodoInput!
) {
  removeTodo(input: $input) {
    deletedTodoId
    user {
      completedCount
      totalCount
      id
    }
  }
}
",
  },
  Object {
    "alias": "67c2bc8aa3185a209d6651b4feb63c04",
    "body": "query appQuery(
  $userId: String
) {
  user(id: $userId) {
    ...TodoApp_user
    id
  }
}

fragment TodoApp_user on User {
  id
  userId
  totalCount
  ...TodoListFooter_user
  ...TodoList_user
}

fragment TodoListFooter_user on User {
  id
  userId
  completedCount
  todos(first: 2147483647) {
    edges {
      node {
        id
        complete
        __typename
      }
      cursor
    }
    pageInfo {
      endCursor
      hasNextPage
    }
  }
  totalCount
}

fragment TodoList_user on User {
  todos(first: 2147483647) {
    edges {
      node {
        id
        complete
        ...Todo_todo
        __typename
      }
      cursor
    }
    pageInfo {
      endCursor
      hasNextPage
    }
  }
  id
  userId
  totalCount
  completedCount
  ...Todo_user
}

fragment Todo_todo on Todo {
  complete
  id
  text
}

fragment Todo_user on User {
  id
  userId
  totalCount
  completedCount
}
",
  },
  Object {
    "alias": "db9904c31d91416f21d45fe3d153884c",
    "body": "mutation MarkAllTodosMutation(
  $input: MarkAllTodosInput!
) {
  markAllTodos(input: $input) {
    changedTodos {
      id
      complete
    }
    user {
      id
      completedCount
    }
  }
}
",
  },
  Object {
    "alias": "2eb8c9941fdb3117fdbc08d15fab62d0",
    "body": "mutation AddTodoMutation(
  $input: AddTodoInput!
) {
  addTodo(input: $input) {
    todoEdge {
      __typename
      cursor
      node {
        complete
        id
        text
      }
    }
    user {
      id
      totalCount
    }
  }
}
",
  },
  Object {
    "alias": "d970fd7dbf118794415dec7324d463e3",
    "body": "mutation RenameTodoMutation(
  $input: RenameTodoInput!
) {
  renameTodo(input: $input) {
    todo {
      id
      text
    }
  }
}
",
  },
  Object {
    "alias": "a49217db31a8be3f4107763b957d5fca",
    "body": "mutation RemoveCompletedTodosMutation(
  $input: RemoveCompletedTodosInput!
) {
  removeCompletedTodos(input: $input) {
    deletedTodoIds
    user {
      completedCount
      totalCount
      id
    }
  }
}
",
  },
  Object {
    "alias": "d7dda774dcfa32fe0d9661e01cac9a4a",
    "body": "mutation ChangeTodoStatusMutation(
  $input: ChangeTodoStatusInput!
) {
  changeTodoStatus(input: $input) {
    todo {
      id
      complete
    }
    user {
      id
      completedCount
    }
  }
}
",
  },
]
`;

exports[`sync operations custom file processing options Adds .graphql to the glob if needed 1`] = `
Array [
  Object {
    "alias": "f7f65309043352183e905e1396e51078",
    "body": "query GetStuff {
  stuff
}
",
    "name": "GetStuff",
  },
]
`;

exports[`sync operations custom file processing options Adds .graphql to the glob if needed 2`] = `
Array [
  Object {
    "alias": "f7f65309043352183e905e1396e51078",
    "body": "query GetStuff {
  stuff
}
",
    "name": "GetStuff",
  },
]
`;

exports[`sync operations custom file processing options Uses a custom hash function if provided 1`] = `
Array [
  Object {
    "alias": "GETSTUFF",
    "body": "query GetStuff {
  stuff
}
",
    "name": "GetStuff",
  },
]
`;

exports[`sync operations generating artifacts without syncing works without a URL 1`] = `
"
    /**
     * Generated by graphql-ruby-client
     *
    */

    /**
     * Map local operation names to persisted keys on the server
     * @return {Object}
     * @private
    */
    var _aliases = {
  \\"GetStuff\\": \\"f7f65309043352183e905e1396e51078\\"
}

    /**
     * The client who synced these operations with the server
     * @return {String}
     * @private
    */
    var _client = \\"test-1\\"

    var OperationStoreClient = {
      /**
       * Build a string for \`params[:operationId]\`
       * @param {String} operationName
       * @return {String} stored operation ID
      */
      getOperationId: function(operationName) {
        return _client + \\"/\\" + OperationStoreClient.getPersistedQueryAlias(operationName)
      },

      /**
       * Fetch a persisted alias from a local operation name
       * @param {String} operationName
       * @return {String} persisted alias
      */
      getPersistedQueryAlias: function(operationName) {
        var persistedAlias = _aliases[operationName]
        if (!persistedAlias) {
          throw new Error(\\"Failed to find persisted alias for operation name: \\" + operationName)
        } else {
          return persistedAlias
        }
      },

      /**
       * Satisfy the Apollo Link API.
       * This link checks for an operation name, and if it's present,
       * sets the HTTP context to _not_ include the query,
       * and instead, include \`extensions.operationId\`.
       * (This is inspired by apollo-link-persisted-queries.)
      */
      apolloLink: function(operation, forward) {
        if (operation.operationName) {
          const operationId = OperationStoreClient.getOperationId(operation.operationName)
          operation.setContext({
            http: {
              includeQuery: false,
              includeExtensions: true,
            }
          })
          operation.extensions.operationId = operationId
        }
        return forward(operation)
      },
      /**
       * Satisfy the Apollo middleware API.
       * Replace the query with an operationId
      */
      apolloMiddleware: {
        applyBatchMiddleware: function(options, next) {
          options.requests.forEach(function(req) {
            // Fetch the persisted alias for this operation
            req.operationId = OperationStoreClient.getOperationId(req.operationName)
            // Remove the now-unused query string
            delete req.query
            return req
          })
          // Continue the request
          next()
        },

        applyMiddleware: function(options, next) {
          var req = options.request
          // Fetch the persisted alias for this operation
          req.operationId = OperationStoreClient.getOperationId(req.operationName)
          // Remove the now-unused query string
          delete req.query
          // Continue the request
          next()
        }
      }
    }

    module.exports = OperationStoreClient
    "
`;

exports[`sync operations verbose Adds debug output 1`] = `
Array [
  Array [
    "[Sync] glob: ",
    "./src/__tests__/documents**/*.graphql*",
  ],
  Array [
    "[Sync] 1 files:",
  ],
  Array [
    "[Sync]   - ./src/__tests__/documents/doc1.graphql",
  ],
  Array [
    "Syncing 1 operations to [1mbogus[0m...",
  ],
  Array [
    "Verbose!",
  ],
  Array [
    "Generating client module in [1msrc/OperationStoreClient.js[0m...",
  ],
  Array [
    "[32mâœ“ Done![0m",
  ],
]
`;
